src/Parser/Monk/Brewmaster/Modules/Spells/TigerPalm.js

```diff
@@ -0,0 +1,138 @@
+import HIT_TYPES from 'Parser/Core/HIT_TYPES';
+import SPELLS from 'common/SPELLS';
+import Analyzer from 'Parser/Core/Analyzer';
+import Combatants from 'Parser/Core/Modules/Combatants';
+import BlackoutCombo from './BlackoutCombo';
+import SharedBrews from '../Core/SharedBrews';
+
+const debug = false;
+const FACE_PALM_AP_RATIO = 1.281;
+const TIGER_PALM_REDUCTION = 1000;
+const FACE_PALM_REDUCTION = 1000;
+
+class TigerPalm extends Analyzer {
+  static dependencies = {
+    combatants: Combatants,
+    boc: BlackoutCombo,
+    brews: SharedBrews,
+  }
+
+  totalCasts = 0;
+  normalHits = 0;
+  bocHits = 0;
+  fpHits = 0;
+  bocFpHits = 0;
+  lastAttackPower = 0;
+
+  cdr = 0;
+  wastedCDR = 0;
+  fpCDR = 0;
+  wastedFpCDR = 0;
+
+  boc_buff_active = false;
+  boc_apply_to_tp = false;
+
+  get facePalmHits() {
+    return this.fpHits + this.bocFpHits;
+  }
+
+  on_byPlayer_applybuff(event) {
+    const spellId = event.ability.guid;
+    if(SPELLS.BLACKOUT_COMBO_BUFF.id === spellId) {
+      this.boc_buff_active = true;
+    }
+  }
+
+  on_byPlayer_refreshbuff(event) {
+    const spellId = event.ability.guid;
+    if(SPELLS.BLACKOUT_COMBO_BUFF.id === spellId) {
+      this.boc_buff_active = true;
+    }
+  }
+
+  on_byPlayer_removebuff(event) {
+    const spellId = event.ability.guid;
+    if(SPELLS.BLACKOUT_COMBO_BUFF.id === spellId) {
+      this.boc_buff_active = false;
+    }
+  }
+
+  on_byPlayer_cast(event) {
+    const spellId = event.ability.guid;
+    if(SPELLS.TIGER_PALM.id === spellId) {
+      this.totalCasts += 1;
+      this.boc_apply_to_tp = this.boc_buff_active;
+    }
+  }
+
+  on_byPlayer_damage(event) {
+    const spellId = event.ability.guid;
+    if(SPELLS.TIGER_PALM.id === spellId) {
+      // OK SO we have a hit, lets reduce the CD by the base amount...
+      const actualReduction = this.brews.reduceCooldown(TIGER_PALM_REDUCTION);
+      this.cdr += actualReduction;
+      this.wastedCDR += TIGER_PALM_REDUCTION - actualReduction;
+
+      if(this._face_palm_proc(event)) {
+        // BUT this is also FACE PALM so we reduce it by an extra amount
+        // too!
+        const actualFpReduction = this.brews.reduceCooldown(FACE_PALM_REDUCTION);
+        this.fpCDR += actualFpReduction;
+        this.wastedFpCDR += FACE_PALM_REDUCTION - actualFpReduction;
+        if(this.boc_apply_to_tp) {
+          this.bocFpHits += 1;
+        } else {
+          this.fpHits += 1;
+        }
+      } else {
+        if(this.boc_apply_to_tp) {
+          this.bocHits += 1;
+        } else {
+          this.normalHits += 1;
+        }
+      }
+    }
+  }
+
+  on_toPlayer_damage(event) {
+    // record the last known AP to estimate the amount of damage a
+    // non-FP TP should do
+    if (event.hasOwnProperty("attackPower") && event.attackPower > 0) {
```

Comment:
`event.hasOwnProperty("attackPower")` should be either just `event.attackPower` or `event.attackPower !== undefined`.